store.rb

```diff
@@ -118,4 +148,39 @@ def get_node(node)
       raise e
     end
   end
+
+  # immediately update cache if node joins/leaves
+  def watch()
+    return if @zk.nil?
+    @root_watcher ||= @zk.register("/", :only => :child) do |event|
+      @log.info "Children added/deleted"
+      reload_instances
+    end
+  end
+
+  def reload_instances()
+    start_seq = 0
+    @mutex.synchronize do
```

Comment:
i'm really confused about the mutex logic here. i don't even get what it's trying to do. a few comments might be helpful?